### DISCUSSION TITLE
How to limit bandwidth for msquic transfer

### DISCUSSION THREAD

**Discussion by chunlei-wu** [microsoft/msquic#3346] [Q&A] [answered]
Hi community,

I want to limit the bandwidth for upload the data to quic server, such as, I just want to send the data limited 1Mbits/s. 
There is any suggestion or example code for it.

Many thanks.

**Comment by nibanks** [comment, marked as answer]
MsQuic doesn't currently support this functionality natively.

@maolson-msft any thoughts here on how/if we should add this support? Right now, it's not particularly easy even for the app to do it themselves.

**Reply by maolson-msft** [reply]
"How?" - probably with a tweak to the pacing logic in CubicCongestionControlGetSendAllowance. Currently it effectively uses a pacing rate of (CongestionWindow / SmoothedRtt); we could add a knob to allow the app to change that to a different target rate. So *send* rate limiting is straightforward; but I'm not sure how we would do *receive* rate limiting, and I wouldn't be surprised if someone down the road asked for that.

"Should?" - I think that, yes, capping the rate falls within the scope of the transport layer. And by adding an explicit API for it, we can hopefully keep apps from doing a shoddy implementation based on limiting buffer sizes (which is what you see today sometimes with winsock apps setting SO_SNDBUF or SO_RCVBUF to limit their throughput). As for the receive side limiting, if we can't come up with a good way to do it, then too bad.

**Reply by nibanks** [reply]
Will connection-wide limiting be enough? What about per-stream (much harder)?

On the receive side, per-stream or per-connection actually wouldn't be any difference in difficulty because we have separate FC windows for each. But, yes, it would be more complicated than the simple proposal you have for the connection-wide send path.

**Reply by maolson-msft** [reply]
One more thought: this is a per-connection cap. I don't have any ideas for a per-app cap.

**Reply by nibanks** [reply]
@chunlei-wu feel free to open an issue to ask for the new feature, explaining your scenario a bit (what exactly you need, and why).

**Reply by chunlei-wu** [reply]
@nibanks @maolson-msft Many thanks your guy, I will create a new feature for this question.
